consent/consent-page/hypr_strategy.go

```diff
@@ -0,0 +1,249 @@
+	}
+
+	switch checkStatus.State[len(checkStatus.State)-1].Value {
+	case "COMPLETED":
+		return nil
+	default:
+		return &MFAError{
+			err:     errors.New("user denied access"),
+			code:    http.StatusUnauthorized,
+			message: "user denied access",
+		}
+	}
+}
+
+func (h *HyprStrategy) SetStorage(r LoginRequest, approved bool) {
+	h.Storage[r] = approved
+}
+
+func (h *HyprStrategy) IsApproved(r LoginRequest) (bool, error) {
+	approved, ok := h.Storage[r]
+	if !ok {
+		return false, nil
+	}
+
+	return approved, nil
+}
+
+func (h *HyprStrategy) startAuthentication(username string) (string, error) {
+	var (
+		endpoint = "/rp/api/oob/client/authentication/requests"
+		resp     *http.Response
+		err      error
+	)
+
+	deviceReq := DeviceAuthenticationRequest{
+		SessionNonce:      GenerateSha256(strconv.Itoa(int(GenerateRandomPin()))),
+		DeviceNonce:       GenerateSha256(strconv.Itoa(int(GenerateRandomPin()))),
+		ServiceHmac:       GenerateSha256(strconv.Itoa(int(GenerateRandomPin()))),
+		ServiceNonce:      GenerateSha256(strconv.Itoa(int(GenerateRandomPin()))),
+		AppID:             h.AppID,
+		NamedUser:         username,
+		Machine:           "WEB",
+		MachineID:         GenerateRandomString(6),
+		AdditionalDetails: nil,
+	}
+
+	if resp, err = h.performRequest(http.MethodPost, fmt.Sprintf("%s%s", h.HostURL, endpoint), deviceReq); err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	data := StartAuthResponse{}
+	if err = json.NewDecoder(resp.Body).Decode(&data); err != nil {
+		return "", err
+	}
+
+	return data.Response.RequestID, nil
+}
+
+func (h *HyprStrategy) poll(requestID string) (*AuthStatusResponse, error) {
+	var (
+		checkStatus  *AuthStatusResponse
+		pollInterval = time.Tick(time.Duration(2) * time.Second)
+		timeout      = time.Tick(time.Duration(120) * time.Second)
+		err          error
+	)
+
+	for {
+		select {
+		case <-timeout:
+			return nil, ErrTimeoutWaitingForUser
+		case <-pollInterval:
+			if checkStatus, err = h.performAuthStatusCheck(requestID); err != nil {
+				return nil, err
+			}
+			for i := range checkStatus.State { // no lint
+				switch checkStatus.State[i].Value {
+				case "COMPLETED", "CANCELED", "FAILED":
+					return checkStatus, nil
+				}
+			}
+		}
+	}
+}
+
+func (h *HyprStrategy) performAuthStatusCheck(requestID string) (*AuthStatusResponse, error) {
+	var (
+		endpoint = fmt.Sprintf("/rp/api/oob/client/authentication/requests/%s", requestID)
+		resp     *http.Response
+		err      error
+	)
+
+	if resp, err = h.performRequest(http.MethodGet, fmt.Sprintf("%s%s", h.HostURL, endpoint), nil); err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	data := AuthStatusResponse{}
+	if err = json.NewDecoder(resp.Body).Decode(&data); err != nil {
+		return nil, err
+	}
+
+	return &data, nil
+}
+
+func (h *HyprStrategy) getUserDevices(username string) (UserDevices, error) {
+	var (
+		endpoint = fmt.Sprintf("/rp/api/oob/client/authentication/%s/%s/devices", "cloudentity", username)
```

Comment:
What does "cloudentity" param represent? Shouldn't that be configurable?

Reply:
yep, missed a hardcode. good catch!